producer/src/secret_manager.rs

```diff
@@ -1,183 +1,274 @@
         };
 
-        info!("ed25519_key_id: {ed25519_key_id:?}, es256_key_id: {es256_key_id:?}, es256k_key_id: {es256k_key_id:?}");
+        // If Stronghold doesn't exist yet, generate new keys
+        if !exists {
+            info!("Generating new keys ...");
+
+            match self.storage_type {
+                StrongholdStorageType::Default => {
+                    let jwk_gen_output = stronghold_storage
+                        .generate(KeyType::new("Ed25519"), JwsAlgorithm::EdDSA)
+                        .await?;
+                    info!(
+                        "Generated new {} key with key ID `{}`",
+                        "Ed25519",
+                        &jwk_gen_output.key_id.as_str()
+                    );
+                    warn!(
+                        "A new key was generated with id `{}`. It is required for all future operations. Provide it using `with_ed25519_key()`. If you do not intend this behavior, remove the generated Stronghold file and use the default key IDs.",
+                        &jwk_gen_output.key_id.as_str()
+                    );
+                    self.ed25519_key_id = Some(jwk_gen_output.key_id);
+                }
+                StrongholdStorageType::Extended => {
+                    let ed25519_key_id =
+                        generate(&stronghold_ext_storage, KeyType::new("Ed25519"), JwsAlgorithm::EdDSA).await?;
+                    self.ed25519_key_id = Some(ed25519_key_id);
+                    let es256_key_id =
+                        generate(&stronghold_ext_storage, KeyType::new("ES256"), JwsAlgorithm::ES256).await?;
+                    self.es256_key_id = Some(es256_key_id);
+                    let es256k_key_id =
+                        generate(&stronghold_ext_storage, KeyType::new("ES256K"), JwsAlgorithm::ES256K).await?;
+                    self.es256k_key_id = Some(es256k_key_id);
+                }
+            }
+        }
 
-        if ed25519_key_id.is_none() && es256_key_id.is_none() && es256k_key_id.is_none() {
-            // TODO: add proper "NoKeysFound" error type
-            return Err(Error::new(ErrorKind::Other, "No keys found"));
+        // Check if specified keys exists
+        match self.storage_type {
+            StrongholdStorageType::Default => {
+                if self.es256_key_id.is_some() || self.es256k_key_id.is_some() {
+                    warn!("Key IDs for ECDSA keys are ignored for Stronghold Storage of type `Default`");
+                }
+                if stronghold_storage
+                    .exists(&self.ed25519_key_id.clone().expect("No key id provided for `Ed25519`"))
+                    .await
+                    .map_err(ProducerError::KeyStorageError)?
+                {
+                    debug!("Key exists: `{}`", &self.ed25519_key_id.clone().unwrap());
+                } else {
+                    return Err(ProducerError::KeyStorageError(SingleStructError::new(KeyNotFound)));
+                }
```

Comment:
Even if now you're 100% certain that there cannot be an edge-case where `self.ed25519_key_id` is not `None` here, I would personally still prefer to return an error here instead of the `expect`. 

If you're really sure you can use `expect` here then maybe consider to put the reference into a variable that you can reuse in line 200 and 204 so you don't need to clone it, e.g:
```rust
                let ed25519_key_id = self.ed25519_key_id.as_ref().expect("No key id provided for `Ed25519`");

                if stronghold_storage
                    .exists(ed25519_key_id)
                    .await
                    .map_err(ProducerError::KeyStorageError)?
                {
                    debug!("Key exists: `{}`", ed25519_key_id);
                } else {
                    return Err(ProducerError::KeyStorageError(SingleStructError::new(KeyNotFound)));
                }
```